### PR TITLE
Key value serdes factories

### LINKED ISSUE #8: How to configure Serializer or Deserializers?
`StringSerializer`, for example has `encoding` configuration. How do you configure it? 

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/serialization/StringSerializer.java#L33

Rather than just one-line class configurations, I would recommend objects

```yaml
valueSerializer:
  class: org.apache.kafka...StringSerializer
  config:
     value.serializer.encoding: "UTF-16"
```

### PR DESCRIPTION
**Pull request by mzamani1**
Fixes #8

### CODE REVIEW

File: src/main/java/io/dropwizard/kafka/deserializer/UUIDDeserializerFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    public String getEncoding() {
+        return encoding;
+    }
+
+    public void setEncoding(final String encoding) {
+        this.encoding = encoding;
+    }
+
+    @JsonIgnore
+    @ValidationMethod(message = "Invalid charset used for StringDeserializerFactory")
```
Comment (cricket007) [inline src/main/java/io/dropwizard/kafka/deserializer/UUIDDeserializerFactory.java]: `for UUIDDeserializerFactory`

File: src/main/java/io/dropwizard/kafka/serializer/SerializerFactory.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package io.dropwizard.kafka.serializer;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.dropwizard.jackson.Discoverable;
+
+import java.util.Map;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+public abstract class SerializerFactory implements Discoverable {
+    public abstract Map<String, Object> build(final boolean isKey);
```
Comment (cricket007) [inline src/main/java/io/dropwizard/kafka/serializer/SerializerFactory.java]: Would it be better to add `public abstract String getSerializerClass()` here?

File: src/main/java/io/dropwizard/kafka/serializer/UUIDSerializerFactory.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public String getEncoding() {
+        return encoding;
+    }
+
+    public void setEncoding(final String encoding) {
+        this.encoding = encoding;
+    }
+
+    @JsonIgnore
+    @ValidationMethod(message = "Invalid charset used for StringSerializerFactory")
```
Comment (cricket007) [inline src/main/java/io/dropwizard/kafka/serializer/UUIDSerializerFactory.java]: `for UUIDSerializerFactory`

File: src/main/java/io/dropwizard/kafka/deserializer/DeserializerFactory.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package io.dropwizard.kafka.deserializer;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.dropwizard.jackson.Discoverable;
+
+import java.util.Map;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+public abstract class DeserializerFactory implements Discoverable {
+    public abstract Map<String, Object> build(final boolean isKey);
```
Comment (cricket007) [inline src/main/java/io/dropwizard/kafka/deserializer/DeserializerFactory.java]: Would it make sense to add `public abstract String getDeserializerClass()`?

**Comment by cricket007** [comment]
Also, would be nice to see a (short) document how to implment, say `confluent-avro` type. 

https://github.com/confluentinc/schema-registry/blob/master/avro-serializer/src/main/java/io/confluent/kafka/serializers/KafkaAvroSerializer.java

File: src/main/java/io/dropwizard/kafka/serializer/SerializerFactory.java (hunk 2 above)
Comment (mzamani1) [reply src/main/java/io/dropwizard/kafka/serializer/SerializerFactory.java]: Seems a bit cleaner, sure!

**Comment by mzamani1** [comment]
@cricket007 Regarding the `KafkaAvroSerializerConfig` question, that's likely what I would do personally. 

I think it could make sense to have additional modules in this repo, that potentially in a `schema` factory, as well as support Avro and other serialization formats, without that being included in the base module where not everyone might want it. Thoughts?

**Comment by cricket007** [comment]
I like the module idea in theory. My only concern is that might not scale well.

Taking Jackson for example - https://github.com/FasterXML/jackson-dataformats-binary

There's enough formats there that I feel like one or few maintainers wouldn't know the intricate details of each format other than just taking relatively simple object types to and from bytes. Maybe that's enough, though, so I'm not really sure.

**Comment by mzamani1** [comment]
@cricket007 I meant creating modules for each of the supported Kafka binary formats, specifically protobuf and avro are the ones I was thinking.

**Comment by cricket007** [comment]
Avro and Protobuf are fairly popular, sure, but why not Thrift or Flatbuffers?

**Comment by mzamani1** [comment]
@cricket007 yeah, fair question, and I think it’d make sense to support any that there’s interest in. I think each module will likely be a tiny amount of code, and most of it being pretty simple. My main objection from having those all being present in a single module is that I have first hand experience with the pain of managing versions of protobuf and Avro, and thus would prefer to limit the exposure of those if possible.

**Pull request merged**